### ISSUE TITLE
Validate fields on loss of focus instead of on submission

### ISSUE THREAD

**Issue by cecilia-donnelly** [OpenTechStrategies/psm#328]
From @kfogel in #325:

>Karl asks: We should talk about how we plan to add immediate validation, so that the user is made aware of such errors as soon as they indicate the field is complete (e.g., by tabbing or clicking away from the field).  [...] Do we agree that this is on the roadmap (even if not for MVP), and that it's something we can talk about?

This enhancement is to do validation of dates, NPIs, etc on click away / loss of focus instead of on page submission.  Some already do this (e.g. NPI), but others, like DOB, don't yet.  Completely this is not part of MVP, but is on the roadmap.

<sub>Confidential to @kfogel: I don't think talking about this is compelling content for a demo.  It'll be good to *show* later, once it's working, but I think describing it will take more time than we have.</sub>

**Comment by brainwane** [comment]
Incidentally, there's a screen currently in the PSM that validates several fields continuously as the user fills them in. The `cms/admin/beginEditAgent?userId=` screen that displays when a service admin creates a new service agent user starts trying to validate the input for the Email and Business Phone fields as soon as the user starts typing in them, and displays an error message (e.g., "Valid phone format is xxx-xxx-xxxx (xxxx), ext. is optional and up to 4 digits.") until the input is satisfactory. So when it comes time to implement this, we can probably reuse whatever approach is used there.

**Comment by cecilia-donnelly** [comment]
Based on conversations with @PaulMorris and @jasonaowen, it sounds like this could be a relatively large task.  Currently the validation takes place in Java, using Drools.  Doing immediate tab-away validation like this would involve moving that to Javascript *or* doing a large number of round trips to the server.  We discussed it with reference to the fact that developers and users can't navigate through forms without filling them in.  This makes it more difficult for us to click around the app to find issues, because we need to do data entry to reach later pages.

We could do all validation on submission of the application (still using Drools), but that presents challenges around how to show errors across multiple pages.

**Comment by slifty** [comment]
I spent some time over the weekend looking into the client side validation question, particularly around whether there are any libraries that allow you to convert drl logic directly to client side.  No promising leads there.  The main suggestion appeared to be to create an API layer for validation, which would allow the client side to run validation logic directly.

Some land mines I'm concerned about as we try to engineer a solution:

1) deep coupling between interface components and specific rules, which would hurt the customization benefits of drools (when a state adds new rules / removes rules, how do we map those new rules to specific pages and field elements.)

2) rules that don't make sense with partially completed forms -- if the current rules have been designed assuming that they only get run after the entire application (or individual form) have been completed, running the full rule set could result in non-ideal user experiences.

3) Violating DRY (and, also, issue {1} again) by building an additional set of validation rules directly into the client side.

TL;DR: I'm worried that the architecture we have is built around the idea of monolithic "all at once" validation, and that trying to make things more granular defeats the purpose of the tools we have selected.  I'll keep looking, and once I have identified a good API-based approach to the task I will pursue the sample form approach.

**Comment by kfogel** [comment]
@slifty's points noted.  We'll have to have an on-mailing-list discussion about how important early validation should be in the PSM (as @cecilia-donnelly and I just agreed in Zulip).

**Comment by cecilia-donnelly** [comment]
Thanks @slifty and @kfogel.  In the meantime, I am moving this out of this milestone.
